### PR TITLE
Bugfixes math // Update GeoIP

### PR DESCRIPTION
**Pull request by miili**
This PR incorporates the bugfixes introduced by https://github.com/oh-my-fish/plugin-weather/pull/28 and https://github.com/oh-my-fish/plugin-weather/pull/24.

* Bugfixes to math
* Bugfix `jq` version check
* Removing DNS IP retrieval, using http instead
* Added Sanity checks for weather response

### CODE REVIEW

**Comment by ericqweinstein** [comment]
Confirming this works locally (at least for me on macOS Mojave, 10.14.3):

```
Weather for Santa Monica, United States

Temperature: 13.88 °C (56.984 °F)
   Humidity: 67%
 Cloudiness: broken clouds
   Pressure: 1019 hpa
       Wind: from S (210°) at 2.6 m/s (9.4 km/h)

5-day forecast
  Temperature: █▇▆▅▄▄▃▆▆▆▄▂▁▁▁▆▇▇▄▂▂▂▂▅▆▆▄▂▂▁▁▆██▅▄▂▁▁█
               07/03  08/03  09/03  10/03  11/03  12/03

Precipitation: ▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅▅
               07/03  08/03  09/03  10/03  11/03  12/03
```

**Comment by abhigenie92** [comment]
Can this be merged into the master, please. :)

**Comment by ahan98** [comment]
Works perfectly after overwriting the original master files!

Here's the output I got BEFORE this fix:

```
Weather for ,

math: Error: Expression is bogus
'null * 3.6'
    ^
math: Error: Missing closing parenthesis
'((null % 360) / 45) + 1'
      ^
math: Error: Expression is bogus
'null - 273.15'
    ^
math: Error: Expression is bogus
' * 1.8 + 32'
  ^
Temperature:  °C ( °F)
   Humidity: null%
 Cloudiness: null
   Pressure: null hpa
       Wind: from  (null°) at null m/s ( km/h)

5-day forecast
jq: error (at <stdin>:1): Cannot iterate over null (null)
  Temperature:
  Temperature:     USAGE:
  Temperature:       spark [-h|--help] VALUE,...
  Temperature:
  Temperature:     EXAMPLES:
  Temperature:       spark 1 5 22 13 53
  Temperature:       ▁▁▃▂█
  Temperature:       spark 0,30,55,80,33,150
  Temperature:       ▁▂▃▄▂█
  Temperature:       echo 9 13 5 17 1 | spark
  Temperature:       ▄▆▂█▁


Precipitation:
Precipitation:     USAGE:
Precipitation:       spark [-h|--help] VALUE,...
Precipitation:
Precipitation:     EXAMPLES:
Precipitation:       spark 1 5 22 13 53
Precipitation:       ▁▁▃▂█
Precipitation:       spark 0,30,55,80,33,150
Precipitation:       ▁▂▃▄▂█
Precipitation:       echo 9 13 5 17 1 | spark
Precipitation:       ▄▆▂█▁
```

And now after:

```
Weather for XXX, United States

Temperature: 9.81 °C (49.658 °F)
   Humidity: 100%
 Cloudiness: light rain
   Pressure: 1020 hpa
       Wind: from SW (240°) at 3.1 m/s (11.2 km/h)

5-day forecast
  Temperature: ▄▃▃▂▁▁▁▁▃▅▅▃▃▂▂▃▄▅▅▄▃▂▁▁▅▆▇▄▂▂▁▂▇█▇▄▂▂▁▂
               21/03  22/03  23/03  24/03  25/03  26/03

Precipitation: ▁▆█▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
               21/03  22/03  23/03  24/03  25/03  26/03
```

(censored city name 😉 )

Please merge into master, and ty @miili !!

**Comment by hemedani** [comment]
please merge into the master branch

**Comment by akrz** [comment]
Can someone merge this?

**Comment by ericqweinstein** [comment]
@scorphus Is this okay to merge? Please let us know if you need any additional info or help! 🙏

**Comment by jenhausu** [comment]
@scorphus please merge this PR, this plugin can't work without this PR 🙏

**Comment by scorphus** [comment]
There're a few issues with this PR. It's just a small part of it that really fixes the issue. It's not just mergeable. Are you willing to fix them, @miili?

**Comment by scorphus** [comment]
Sorry, closed by accident. Will review and request quite a few changes.

File: functions/weather.fish [hunk 1]
```diff
@@ -8,7 +8,7 @@ function weather -d "Displays weather info"
     return 1
   else
     set -l jq_version (jq --version 2>&1 | tr -dC '[:digit:].')
-    if math "$jq_version<1.5" > /dev/null
+    if test (echo "$jq_version < 1.5" | bc ) -eq 1 > /dev/null
```
Comment (scorphus) [inline functions/weather.fish]: Should use just `test` instead.
```suggestion
    if test "$jq_version" -lt 1.5
```

File: functions/weather.fish [hunk 2]
```diff
@@ -49,15 +49,17 @@ function weather -d "Displays weather info"
   set wind_deg (echo $json | jq '.wind.deg')
 
   # convert m/s to km/h
-  set wind_speed_kmh ( math -s 1 $wind_speed \* 3.6 )
-  set  wind_gust_kmh ( math -s 1  $wind_gust \* 3.6 )
+  set wind_speed_kmh ( math -s 1 "$wind_speed * 3.6" )
+  if not test $wind_gust = null
```
Comment (scorphus) [inline functions/weather.fish]: Add only the `if` statement and be sure to use `test` correctly:
```suggestion
  if test "$wind_gust" != "null" -a -n "$wind_gust"
```

File: functions/weather.fish [hunk 3]
```diff
@@ -49,15 +49,17 @@ function weather -d "Displays weather info"
...
+  end
 
   # If we want miles per hour
-  #set wind_speed_mph ( math -s 1 $wind_speed \* 2.23694 )
+  #set wind_speed_mph ( math -s 1 "$wind_speed * 2.23694" )
 
   # Get the cardinal direction from the heading
   set directions N NE E SE S SW W NW N
-  set wind_dir $directions[(math "(($wind_deg % 360) / 45) + 1")]
+  set wind_dir $directions[(math -s0 "(($wind_deg % 360) / 45) + 1")]
```
Comment (scorphus) [inline functions/weather.fish]: Don't truncate. Round instead:

```suggestion
  set wind_dir $directions[(math "round(($wind_deg % 360) / 45) + 1")]
```

This:

```
echo $directions[(math "round((89 % 360) / 45) + 1")]
E
```

Is a bit more accurate than this:

```
echo $directions[(math -s0 "((89 % 360) / 45) + 1")]
NE
```

File: functions/weather.location.fish [hunk 4]
```diff
@@ -6,36 +6,34 @@ function weather.location -d "Get the current geographic location"
   end
 
   # Fetch location data based on our IP
-  if not set geoip_data (weather.fetch "http://geoip.nekudo.com/api/$ip")
+  if not set geoip_data (curl -s "https://ipapi.com/ip_api.php?ip=$ip")
```
Comment (scorphus) [inline functions/weather.location.fish]: Please make sure to use `weather.fetch`, it's there for a reason.
```suggestion
  if not set geoip_data (weather.fetch "https://ipapi.com/ip_api.php?ip=$ip")
```

File: functions/weather.location.fish [hunk 5]
```diff
@@ -6,36 +6,34 @@ function weather.location -d "Get the current geographic location"
...
     echo "Unable to query GeoIP data; please try again later."
     return 1
   end
-
+
```
Comment (scorphus) [inline functions/weather.location.fish]: Please remove this from the diff, it's not part of the fix.

File: functions/weather.location.fish [hunk 6]
```diff
@@ -6,36 +6,34 @@ function weather.location -d "Get the current geographic location"
...
-  end
+  # if set -q __weather_system_dns
+  #  if not set ip (dig +short myip.opendns.com)
+  #    return 1
+  #  end
+  # else
+  #  if not set ip (dig +short myip.opendns.com @resolver1.opendns.com)
+  #    return 1
+  #  end
+  # end
```
Comment (scorphus) [inline functions/weather.location.fish]: Please undo all this.

File: functions/weather.location.fish [hunk 7]
```diff
@@ -6,36 +6,34 @@ function weather.location -d "Get the current geographic location"
...
 
   # Attempt to get our external IP using a web service.
-  if test -z $ip
-    if not set ip (weather.fetch "http://ipecho.net/plain")
-      return 1
-    end
+  if not set ip (weather.fetch "http://ipecho.net/plain")
+    return 1
```
Comment (scorphus) [inline functions/weather.location.fish]: Please undo all this.

File: functions/weather.location.fish [hunk 8]
```diff
@@ -6,36 +6,34 @@ function weather.location -d "Get the current geographic location"
...
   end
-
+
```
Comment (scorphus) [inline functions/weather.location.fish]: Please remove this from the diff, it's not part of the fix.

**Review by scorphus — CHANGES_REQUESTED** [review]
Thanks for your contribution! Please make the requested changes.

**Comment by ericqweinstein** [comment]
@miili Do you have time/bandwidth to make the changes @scorphus requested, or would you like me to pick this one up?

File: functions/weather.fish [hunk 9]
```diff
@@ -49,15 +49,17 @@ function weather -d "Displays weather info"
...
-  set wind_speed_kmh ( math -s 1 $wind_speed \* 3.6 )
-  set  wind_gust_kmh ( math -s 1  $wind_gust \* 3.6 )
+  set wind_speed_kmh ( math -s 1 "$wind_speed * 3.6" )
+  if not test $wind_gust = null
+    set wind_gust_kmh ( math -s 1 "$wind_gust * 3.6" )
+  end
 
   # If we want miles per hour
-  #set wind_speed_mph ( math -s 1 $wind_speed \* 2.23694 )
+  #set wind_speed_mph ( math -s 1 "$wind_speed * 2.23694" )
```
Comment (scorphus) [inline functions/weather.fish]: Also, please remove this from the diff, it's not part of the fix, besides being dead code – which is a subject of another commit.

File: functions/weather.fish (hunk 1 above)
Comment (miili) [reply functions/weather.fish]: No, does not work.

File: functions/weather.location.fish (hunk 5 above)
Comment (miili) [reply functions/weather.location.fish]: This structures the code, visually.

File: functions/weather.location.fish (hunk 6 above)
Comment (miili) [reply functions/weather.location.fish]: This is broken, it needs to be removed. Commented as a courtesy.

File: functions/weather.location.fish (hunk 8 above)
Comment (miili) [reply functions/weather.location.fish]: This structures the code.

**Comment by miili** [comment]
> @miili Do you have time/bandwidth to make the changes @scorphus requested, or would you like me to pick this one up?

Hey @ericqweinstein, please go ahead!
Some of the requested changes will break the plugin on my system, please mind my comments.

File: functions/weather.location.fish (hunk 6 above)
Comment (jenhausu) [reply functions/weather.location.fish]: I agree with Scorphus. This is a bad smell for code. We should either figure it out what's going on or just delete it, after all, anyone can find this code in the history.

I'm not familiar with shell development. If you can tell me how to run this code locally, maybe I can help.

Comment (scorphus) [reply functions/weather.location.fish]: Not only it isn't broken, getting our external IP through `dig` and OpenDNS is way faster:

```
»»»» time dig +short myip.opendns.com @resolver1.opendns.com
109.41.193.125
        0.05 real         0.00 user         0.00 sys
```

ipecho.net is there as a fallback:

```
»»»» time curl -s http://ipecho.net/plain
123.12.23.123
        0.27 real         0.00 user         0.00 sys
```

File: functions/weather.location.fish (hunk 8 above)
Comment (scorphus) [reply functions/weather.location.fish]: Save some cases (e.g. Markdown) trailing spaces are unnecessary and just polute the code/markup.

**Comment by scorphus** [comment]
As wisely raised by @derekstavis in another question, we need to make sure all changes here are backward compatible with Fish 2.x, so the plugin still works for users or on distros that still didn't upgrade.

**Comment by rousseldenis** [comment]
Hi all, what's the status of this ?

**Comment by ericqweinstein** [comment]
@scorphus Since I would have had to be added as a collaborator on @miili's fork, I created a new one (#32)—please let me know if you'd prefer I take a different approach. Happy to make any changes you need.

**Comment by scorphus** [comment]
If you don't mind, I'd rather squash and merge #29, for authorship reasons. Also, #32 does not add anything on top of that.

Then I'll fix a few issues with it and create another pull request for review.

**Comment by scorphus** [comment]
Also, let's not forget this is partially a dupe of #28. So I'll rebase these changes on top of those.

**Comment by scorphus** [comment]
Please review and comment.

**Review by sagebind — APPROVED** [review]

**Review by scorphus — APPROVED** [review]

**Comment by ericqweinstein** [comment]
🎉

**Pull request merged**